Refactor Unit Tests

dylandesrosier opened the pull request:
- Tests have been refactored
- pulled `_applyLiquidityFraction` logic into a function so we can directly test it and rewrote the math to avoid overflows
- `UFixed32x9` -> `UFixed32x4`
    - 9 decimal places of precision for the Swap Multiplier and Liquidity Fraction was unnecessarily large and causes overflows. When applying the Liquidity Fraction we would need 32 bits to scale the `amountIn` in some situations (really low LF with really high amountIn). This _could_ be handled by restricting the amountIn; however, 4 decimal places is sufficient for these numbers and doesn't require any restrictions to ensure space in the variables.

Commit:
Format

Commit:
WIP

Commit:
Full test coverage

PierrickGT commented on src/LiquidationPair.sol:
```diff
@@ -42,18 +47,39 @@ contract LiquidationPair {
       "LiquidationPair/liquidity-fraction-less-than-one"
     );
+    require(
+      uint256(_virtualReserveIn) * _virtualReserveOut >= _minK,
```
Why not cast everything to `uint256`?
```suggestion
      uint256(_virtualReserveIn * _virtualReserveOut) >= _minK,
```

dylandesrosier replied:
<img width="730" alt="Screenshot 2023-04-06 at 3 36 59 PM" src="https://user-images.githubusercontent.com/13701258/230480497-6d71d901-f585-420e-8f16-33a801d68b4a.png">

PierrickGT commented on src/LiquidationPair.sol:
```diff
@@ -42,18 +47,39 @@ contract LiquidationPair {
+      "LiquidationPair/virtual-reserves-greater-than-min-k"
+    );
+    require(_minK > 0, "LiquidationPair/min-k-greater-than-zero");
+    require(_virtualReserveIn <= type(uint112).max, "LiquidationPair/virtual-reserve-in-too-large");
+    require(
+      _virtualReserveOut <= type(uint112).max,
+      "LiquidationPair/virtual-reserve-out-too-large"
+    );
```
Would make more sense to directly cast `_virtualReserveIn` and `_virtualReserveOut` to `uint112` in the constructor instead of having these two requires.

dylandesrosier replied:
We want them to be `uint128` but with a max value of `type(uint112).max` on initialization so the virtual reserves can expand during the `virtualBuyback` then shrink back down during the actual swap the user is executing.

PierrickGT commented on src/LiquidationPair.sol:
```diff
@@ -42,18 +47,39 @@ contract LiquidationPair {
 
   /* ============ External Function ============ */
 
+  function maxAmountIn() external returns (uint256) {
```
Useless function?

PierrickGT commented on src/libraries/LiquidatorLib.sol:
```diff
@@ -32,34 +31,107 @@ library LiquidatorLib {
     uint256 numerator = amountOut0 * reserve1;
     uint256 denominator = uint256(reserve0) - amountOut0;
-    amountIn1 = (numerator / denominator);
+    amountIn1 = (numerator / denominator) + 1;
```
Why we offset by 1 here?

dylandesrosier replied:
<https://github.com/Uniswap/v2-periphery/blob/0335e8f7e1bd1e8d8329fd300aea2ef2f36dd19f/contracts/libraries/UniswapV2Library.sol#L58>
To handle rounding when computing the amount in for a given amount out. It ensures that truncation impact always goes in favour of the Liquidator - the user contributes more than necessary to get the requested `amountOut` rather than contributing less than necessary.

Commit:
Shrink to 4 decimals to ensure liquidity fraction scaling fits

Commit:
Fix test

Commit:
Actually fix test

Commit:
100% coverage

Commit:
Fix Overflow when applying liquidity fraction

- Add a check for if the updated virtual reserves fit into uint112 as well
as a check for if the new K will fit into a uint256.
- Add updated virtual reserve data to events

Commit:
Document contracts
